lib/ansible/modules/network/dcnm/dcnm_vrf.py

```diff
@@ -0,0 +1,284 @@
+    payload['vrfId'] = module.params['vrf_id']
+
+    json_data = json.dumps(payload)
+
+    logit(json_data)
+
+    return json_data
+
+
+def vrf_attach_payload(module):
+
+    payload_list = list()
+    payload = dict()
+    payload['vrfName'] = module.params['vrf_name']
+    payload['lanAttachList'] = list()
+
+    for serials in module.params['serial_numbers_vlans']:
+        for serial, vlan in serials.items():
+            device_payload = dict()
+            device_payload['fabric'] = module.params['fabric']
+            device_payload['vrfName'] = module.params['vrf_name']
+            device_payload['serialNumber'] = serial
+            device_payload['vlan'] = vlan
+            device_payload['deployment'] = module.params['deployment']
+            device_payload['extensionValues'] = ""
+            device_payload['instanceValues'] = ""
+            device_payload['freeformConfig'] = ""
+
+        payload['lanAttachList'].append(device_payload)
+
+    payload_list.append(payload)
+    json_data = json.dumps(payload_list)
+
+    return json_data
+
+
+def vrf_deploy_payload(module):
+
+    payload = dict()
+    payload['vrfNames'] = module.params['vrf_name']
+
+    json_data = json.dumps(payload)
+
+    return json_data
+
+
+def main():
+    """ main entry point for module execution
+    """
+
+    element_spec = dict(
+        action=dict(required=True, choices=['create', 'attach', 'deploy']),
+        fabric=dict(required=True, type='str'),
+        vrf_name=dict(required=True, type='str'),
+        vrf_template=dict(default='Default_VRF_Universal', type='str'),
+        vrf_extension_template=dict(default='Default_VRF_Extension_Universal', type='str'),
+        vrf_template_config=dict(default='', type='str'),
+        vrf_id=dict(type=int),
+        serial_numbers_vlans=dict(type='list'),
+        deployment=dict(default=False, type='bool')
+    )
+
+    required_one_of = [['vrf_id', 'vrf_name']]
+    mutually_exclusive = [['vrf_id'],
+                          ['serial_numbers_vlans']]
+
+    module = AnsibleModule(argument_spec=element_spec,
+                           required_one_of=required_one_of,
+                           mutually_exclusive=mutually_exclusive,
+                           supports_check_mode=True)
+
+    result = dict(
+        changed=False,
+        response=dict()
+    )
+
+    action = module.params['action']
+    fabric = module.params['fabric']
+
+    method = 'POST'
+    path = '/rest/top-down/fabrics/{}/vrfs'.format(fabric)
+
+    conn = Connection(module._socket_path)
+
+    if action == 'create':
+        if not check_vrf_exists(module, conn):
+            json_data = vrf_create_payload(module)
+            result['response'] = conn.send_request(method, path, json_data)
+            result['changed'] = True
+        else:
+            result['changed'] = False
+    elif action == 'attach':
+        json_data = vrf_attach_payload(module)
```

Comment:
Is there a way we can query to see if something is attached already and not try attaching again if so?

Reply:
Yeah, we do have an API to query the current vrf attachments. The attach VRF API itself does not error if we try the same attachment again. Wondering whats the value add if we put in a check before calling attach API.